### ISSUE TITLE
Enable profiling optional tensor types in Profiler

### ISSUE THREAD

**Issue by Krovatkin** [pytorch/pytorch#67753]
## 🚀 Feature

Enable profiling optional tensor types in Profiler.
Note, this feature would be **much easier** to implement after https://github.com/pytorch/pytorch/issues/67751

## Motivation

This would allow NVFuser to specialize a number of ops such as

```
aten::instance_norm                                                                   
aten::_batch_norm_impl_index                                                        
aten::native_batch_norm                                                             
aten::batch_norm                                                                    
aten::_batch_norm_impl_index_backward                                               
aten::native_layer_norm                                                             
aten::layer_norm                                                                    
aten::native_layer_norm_backward                                                    
aten::linear                                                                        
prim::add_optional                                                                  
aten::_batch_norm_impl_index                                                        
aten::native_batch_norm                                                             
aten::batch_norm                                                                    
aten::instance_norm                                                                 
aten::native_layer_norm                                                             
aten::layer_norm                                                                    
aten::_batch_norm_impl_index_backward                                               
aten::native_layer_norm_backward
```

see https://github.com/pytorch/pytorch/issues/67753#issuecomment-959730294

## Pitch

1. We assume that https://github.com/pytorch/pytorch/issues/67751 and we are merging the profiled types directly into `profiled_type`.
2. Relax a check https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L289 to also allow profiling of optional types. You could use `i->type()->isSubtypeOf(OptionalType::ofTensor())` or something more efficient.
3. add a new opcode `prim::profile_opt` in `interned_strings.h` **CONTEXT**: we have two options here: we can either reuse `prim::profile` for also profiling optional tensors or we could create a brand new `profile_opt`. The main advantage of the latter would be that we don't have to modify all the uses `prim::profile` as they expect non-optional tensor types. We could slowly start enabling it for cases we care about (autodiff, fuser, etc). It would be completely opt-in. If a pass doesn't know about the existence of `prim::profile_opt` but can handle `prim::profile`, it should still work.
4. Create a new ProfileOptOp class similar to https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L1454
5. In https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L86 create either `
6. In `https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L168` before we make any updates to `profiled_type` add the logic that does the following:
  a. Checks if the value is of an optional type and if true
     i if (ivalue.isNone) then set an integer attribute val->node()->i(Symbol::attr("seen_none")) to 1
     ii otherwise fall-through
7. make a sweep over all instances `prim::profile_ivalue`, it needs to be added to IR printer, side effects, and alias analysis and it should also be removed by any pass that removes all types of profiling nodes.
8. add tests that test the following scenarios:
   a. a tensor is never none
   b. a tensor is always none
   c. a tensor was none in one iteration and defined in the second iteration
   check that prim::profile nodes are annotated with `seen_none` and `profiled_type`
9 any pass that cares about optional tensors should look for `prim::profile_opt` which should have two attributes: `profiled_type` and `seen_none`

## Additional context

1. Learning how to build PyTorch and run tests (1-3 days)
2. Reading `profiling_record.cpp`/`profiling_record.h` (1-2 days)
3. Adding support to profile optional tensors `profiling_record.cpp` and introducing `ProfileOptOp` (2-4 days)
4. Making a sweep (1-3 days)
5. Adding tests (1-3 days)
6. Running CI (1-3 days)

Before working on this task, it's recommended to acquire some basic understanding of PE flow. This can be done by writing a simple model (2-3 ops) or using one of our tests in `test_jit.py` or `test_jit_fuser_te.py` and running it multiple times (1st time for profiling, 2nd time for optimization) and inspecting the output of `export PYTORCH_JIT_LOG_LEVEL=>>profiling_graph_executor_impl`

**Comment by Krovatkin** [comment]
@jjsjann123 could you please remind me which ops use optional tensors, so I can fill in the motivation section.

@jjsjann123  @eellison wdyt ? 

fyi: @penguinwu

**Comment by jjsjann123** [comment]
> @jjsjann123 could you please remind me which ops use optional tensors, so I can fill in the motivation section.

Ops with optional tensor that we currently interested in fuser are:
```
aten::instance_norm                                                                   
aten::_batch_norm_impl_index                                                        
aten::native_batch_norm                                                             
aten::batch_norm                                                                    
aten::_batch_norm_impl_index_backward                                               
aten::native_layer_norm                                                             
aten::layer_norm                                                                    
aten::native_layer_norm_backward                                                    
aten::linear                                                                        
prim::add_optional                                                                  
aten::_batch_norm_impl_index                                                        
aten::native_batch_norm                                                             
aten::batch_norm                                                                    
aten::instance_norm                                                                 
aten::native_layer_norm                                                             
aten::layer_norm                                                                    
aten::_batch_norm_impl_index_backward                                               
aten::native_layer_norm_backward
```

The list is longer if we look at all aten ops.

**Comment by jjsjann123** [comment]
No strong opinion, but leaning towards reuse `prim::profile`.

One thing I want to note is, if we go with option 1 to reuse `prim::profile` for optional tensors, wee need to refactor its implementation to properly profile `none`. If `seen_none` is set, we would want to explicitly list that. Right now `profiled_type` is initialized as `TensorType::get()` ( not sure if the other refactor work is going to change this #67751 ), we need to distinguish `not seen` from `seen none` if we want to reuse the node.

My intuition is, reuse `prim::profile` should not be too hard if we simply default `profiled_type` to `TensorType::get()` and mark `seen_none` in a separate attribute. Might need to refactor some code that restores the output tensor type to be `Optional[Tensor]` when inputs to `prim::profile` is an `Optional[Tensor]`. Passes that doesn't handle `seen_none` would simply be missing out some optimization opportunities, but they should still be functional.

**Comment by jjsjann123** [comment]
Another consideration of reusing `prim::profile`: If an operation's signature takes `Optional[Tensor]` as an argument, we can actually feed both `Tensor` and `Optional[Tensor]` to that op.

E.g. If we look at `linear(Tensor input, Tensor weight, Tensor? bias=None) -> Tensor`, the op takes both `Tensor` and `Optional[Tensor]` as its third argument.

```
   def t(x : torch.Tensor, w : torch.Tensor, b : Optional[torch.Tensor]):                                                
   #def t(x : torch.Tensor, w : torch.Tensor, b : torch.Tensor):                                                                                                                           
       o = torch.nn.functional.linear(x, w, b)                                                                           
       return o
```

So if we simply reuse `prim::profile`, in cases where optional tensor are profiled, we don't need to change any existing code and it would work out of the box.

In the case where `seen_none` is passed there. We'll just go through fallback path and not fusing the tensor (nvfuser specific). I can imagine other consumers go into a loop of bailout in attempt to re-profile `Optional[Tensor]` <- might be a little messy there. You guys know much more about other consumers, maybe it's a big hassle?

**Comment by wukong1992** [comment]
@Krovatkin The explanation is so cryptic, I took three days to read, still can't understand, can you give some examples. Thank you![profile](https://github.com/pytorch/pytorch/blob/834bd3134eab322151a769589f6fb47c6c888342/torch/csrc/jit/runtime/profiling_record.h#L15)

**Comment by Krovatkin** [comment]
@wukong1992 hey, sorry for the delayed reply. 

I'd suggest to try a very simple model such as

```python
import torch

# run this with `export PYTORCH_JIT_LOG_LEVEL=">>profiling_graph_executor_impl" && python test_model.py `
@torch.jit.script
def model(x, y, z, w):
    return (x + y) * z - w
    
    
a = torch.rand(2, 2, device = 'cuda')
b = torch.rand(2, 2, device = 'cuda')
c = torch.rand(2, 2, device = 'cuda')
d = torch.rand(2, 2, device = 'cuda')

_ = model(a, b, c, d)
_ = model(a, b, c, d)
```

If you search for `Profiled Graph` in the output you will see something similar to : 

```
[DUMP profiling_graph_executor_impl.cpp:678] Profiled Graph:
[DUMP profiling_graph_executor_impl.cpp:678] graph(%x.1 : Tensor,
[DUMP profiling_graph_executor_impl.cpp:678]       %y.1 : Tensor,
[DUMP profiling_graph_executor_impl.cpp:678]       %z.1 : Tensor,
[DUMP profiling_graph_executor_impl.cpp:678]       %w.1 : Tensor):
[DUMP profiling_graph_executor_impl.cpp:678]   %4 : int = prim::Constant[value=1]()
[DUMP profiling_graph_executor_impl.cpp:678]   %8 : Tensor = prim::profile[profiled_type=Tensor](%x.1)
[DUMP profiling_graph_executor_impl.cpp:678]   %9 : Tensor = prim::profile[profiled_type=Tensor](%y.1)
[DUMP profiling_graph_executor_impl.cpp:678]   %5 : Tensor = aten::add(%8, %9, %4) # test_kong.py:6:12
[DUMP profiling_graph_executor_impl.cpp:678]   %10 : Tensor = prim::profile[profiled_type=Tensor](%5)
[DUMP profiling_graph_executor_impl.cpp:678]   %11 : Tensor = prim::profile[profiled_type=Tensor](%z.1)
[DUMP profiling_graph_executor_impl.cpp:678]   %6 : Tensor = aten::mul(%10, %11) # test_kong.py:6:12
[DUMP profiling_graph_executor_impl.cpp:678]   %12 : Tensor = prim::profile[profiled_type=Tensor](%6)
[DUMP profiling_graph_executor_impl.cpp:678]   %13 : Tensor = prim::profile[profiled_type=Tensor](%w.1)
[DUMP profiling_graph_executor_impl.cpp:678]   %7 : Tensor = aten::sub(%12, %13, %4) # test_kong.py:6:12
[DUMP profiling_graph_executor_impl.cpp:678]   %14 : Tensor = prim::profile[profiled_type=Tensor](%7)
[DUMP profiling_graph_executor_impl.cpp:678]    = prim::profile()
[DUMP profiling_graph_executor_impl.cpp:678]   return (%14)
```

We take the original model and insert special `prim::profile` nodes that will remember the data type, shape, stride properties about its input tensor. 

If you scroll a little bit down in the output **after Profiled Graph**, until `[DEBUG profiling_graph_executor_impl.cpp:503] Before runProfilingOptimizations`

you will see that we after running the model once, we annotated each `prim::profile` with `profiled_type` . **Note, we only profiled tensor values/outputs**

```
[DEBUG profiling_graph_executor_impl.cpp:503] graph(%x.1 : Tensor,
[DEBUG profiling_graph_executor_impl.cpp:503]       %y.1 : Tensor,
[DEBUG profiling_graph_executor_impl.cpp:503]       %z.1 : Tensor,
[DEBUG profiling_graph_executor_impl.cpp:503]       %w.1 : Tensor):
[DEBUG profiling_graph_executor_impl.cpp:503]   %4 : int = prim::Constant[value=1]()
[DEBUG profiling_graph_executor_impl.cpp:503]   %5 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cuda:0)](%x.1)
[DEBUG profiling_graph_executor_impl.cpp:503]   %6 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cuda:0)](%y.1)
[DEBUG profiling_graph_executor_impl.cpp:503]   %7 : Tensor = aten::add(%5, %6, %4) # test_kong.py:6:12
[DEBUG profiling_graph_executor_impl.cpp:503]   %8 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cuda:0)](%7)
[DEBUG profiling_graph_executor_impl.cpp:503]   %9 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cuda:0)](%z.1)
[DEBUG profiling_graph_executor_impl.cpp:503]   %10 : Tensor = aten::mul(%8, %9) # test_kong.py:6:12
[DEBUG profiling_graph_executor_impl.cpp:503]   %11 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cuda:0)](%10)
[DEBUG profiling_graph_executor_impl.cpp:503]   %12 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cuda:0)](%w.1)
[DEBUG profiling_graph_executor_impl.cpp:503]   %13 : Tensor = aten::sub(%11, %12, %4) # test_kong.py:6:12
[DEBUG profiling_graph_executor_impl.cpp:503]   %14 : Tensor = prim::profile[profiled_type=Float(2, 2, strides=[2, 1], requires_grad=0, device=cuda:0)](%13)
```

If you try to run a model that contains `torch.nn.functional.layer_norm` op, you would see that we wouldn't insert anything for `bias` because it's `Optional[Tensor]`

This is exactly what this issue needs to address.

Please let me know if you have any more questions and if you would like to work on this issue.

**Comment by Krovatkin** [comment]
@wukong1992 if you decide to work on this, I'd suggest you base your work on https://github.com/pytorch/pytorch/pull/69096 since it greatly simplifies our profiler.

fyi: @jjsjann123  we are one step there !! yay!

**Comment by wukong1992** [comment]
@Krovatkin Thanks for the detailed explanation, I understand it overall, and saved me a lot of time, I will take this issue. But there is still a little doubt about: https://github.com/pytorch/pytorch/blob/b2e79ed5ecabcf4be299dc2ed085223ab5c22fd7/torch/csrc/jit/runtime/profiling_record.h#L36-L59 . Can you give me some examples>

**Comment by Krovatkin** [comment]
@wukong1992 we removed all the code this comment refers to, so Profiler doesn't need to create ShapeSymbols anymore. At most, you might need to know about this function: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L150 though I'm not 100% sure, this knowledge is needed to finish this task.

**Comment by davidberard98** [comment]
Should be fixed by https://github.com/pytorch/pytorch/pull/70532
